server/server.go

```diff
@@ -427,9 +427,17 @@ func (s *Server) ClusterName() string {
 // setClusterName will update the cluster name for this server.
 func (s *Server) setClusterName(name string) {
 	s.mu.Lock()
+	var resetCh chan struct{}
+	if s.sys != nil && s.info.Cluster != name {
+		// can't hold the lock as go routine reading it may be waiting for lock as well
+		resetCh = s.sys.resetCh
+	}
 	s.info.Cluster = name
 	s.routeInfo.Cluster = name
 	s.mu.Unlock()
+	if resetCh != nil {
+		resetCh <- struct{}{}
```

Comment:
What is this for?

Reply:
causes internalSendLoop to pick up the changed cluster name

Reply:
Can this block?

Reply:
Not for long. 

This is a blocking channel in order to have certain guarantees with respect to message processing. Meaning, all messages being processed by internalSendLoop after sending to resetCh returned are guaranteed to use the new value. (avoid flappers in other situations we send to resetCh)

The most this waits is for is internalSendLoop to finish processing the current message.
Since this send is not really supposed to happen and on top of that is only done when smth actually changes I wouldn't worry.

Reply:
So on the other side why not just look for a change? Do we cache the whole info block and need this to detect a change when we could just do a quick lock and update it?

Reply:
On the other side, while holding the server lock, we copy values that rarely/never change onto the heap such that the server lock won't be needed on every message.
Among these are: trace level/system account/cluster name/hostname/servername.

I basically noticed that reset was missing and added it.

Reply:
ok, I get nervous when we use channels, not saying we shouldn't but when something should never block and we have a channel push in there makes me nervous. You can make the push non-blocking as well, maybe that is the better answer here? @kozlovic what do you think? Something to keep pushing on or good as is?

There is a bug I need to fix that I saw doing changes to the go client as well around cluster naming and selection.

Reply:
I already reviewed when he introduced the resetCh for config reload of authorization and it was fine. I think here it is fine too, but obviously we need to make sure that nothing in the calling stack has a lock that the internal sendLoop may be waiting on. I don't think that is the case since this function (setClusterName) gets/releases the server lock, so if the calling stack was holding any kind of other lock (say client, account, etc..) we would have already an issue there.